[iOS] Mitigate crashes under WebPage::requestDocumentEditingContext() when handling spatial requests

whsieh opened the pull request:
#### 1f6e2d556f9ba93080f6b215e9e2a1e0096daab0
<pre>
[iOS] Mitigate crashes under WebPage::requestDocumentEditingContext() when handling spatial requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=289910">https://bugs.webkit.org/show_bug.cgi?id=289910</a>
<a href="https://rdar.apple.com/146901044">rdar://146901044</a>

Reviewed by Abrar Rahman Protyasha and Tim Horton.

When handling a spatial document editing context request, logic introduced in 289055@main and
290710@main will attempt to clamp the context range (`rangeOfInterest`) to the extents of the
current editable root (if one exists), or the farthest positions from the current selection
(marching outward by line granularity), for which the visible position remains within the requested
spatial rect. After the latter change, we now use `std::clamp` instead of manually clamping the
start/end `rangeOfInterest` positions by directly using the `&lt;` and `&gt;` operators.

The change to use `std::clamp` wasn&apos;t intended to change behavior, but it did introduce a subtle
bug — if the `startPositionNearSelection` and `endPositionNearSelection` passed into `std::clamp`
are such that `startPositionNearSelection &gt; endPositionNearSelection`, we end up crashing with a
`brk #0x1`. Although I was unable to reproduce this locally, I was able to induce an identical crash
by swapping `startPositionNearSelection` and `endPositionNearSelection`, which forces the starting
position to be greater than the ending position.

To fix the crash, we make the clamping logic robust against this scenario by only clamping if the
start position is less than or equal to the ending position.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Canonical link: <a href="https://commits.webkit.org/292286@main">https://commits.webkit.org/292286@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e9a0e59331b2f108f5a360ade7e1e94fa2b3c942

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15068 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/4995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100514 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45971 "Build is in progress. Recent messages:Running configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15353 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23499 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/45971 "Build is in progress. Recent messages:Running configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98469 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11505 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86223 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11211 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/4995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45307 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88134 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81405 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102550 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22516 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16446 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81858 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22768 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82233 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81209 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20347 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25785 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3248 "Exiting early after 60 failures. 56261 tests run. 60 failures") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15843 "Build is in progress. Recent messages:Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22484 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27638 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22143 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25619 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23885 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Running checkout-pull-request; Compiled WebKit (warnings)") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
I thought you would have gone with a solution using `std::minmax()`, but I suppose `rangeOfInterest` doesn't make sense when `startPosition > endPosition`.

whsieh committed:
[iOS] Mitigate crashes under WebPage::requestDocumentEditingContext() when handling spatial requests
https://bugs.webkit.org/show_bug.cgi?id=289910
rdar://146901044

Reviewed by Abrar Rahman Protyasha and Tim Horton.

When handling a spatial document editing context request, logic introduced in 289055@main and
290710@main will attempt to clamp the context range (`rangeOfInterest`) to the extents of the
current editable root (if one exists), or the farthest positions from the current selection
(marching outward by line granularity), for which the visible position remains within the requested
spatial rect. After the latter change, we now use `std::clamp` instead of manually clamping the
start/end `rangeOfInterest` positions by directly using the `<` and `>` operators.

The change to use `std::clamp` wasn't intended to change behavior, but it did introduce a subtle
bug — if the `startPositionNearSelection` and `endPositionNearSelection` passed into `std::clamp`
are such that `startPositionNearSelection > endPositionNearSelection`, we end up crashing with a
`brk #0x1`. Although I was unable to reproduce this locally, I was able to induce an identical crash
by swapping `startPositionNearSelection` and `endPositionNearSelection`, which forces the starting
position to be greater than the ending position.

To fix the crash, we make the clamping logic robust against this scenario by only clamping if the
start position is less than or equal to the ending position.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Canonical link: https://commits.webkit.org/292286@main
